components/kyma-operator/pkg/kymasources/sourcegetter.go

```diff
@@ -0,0 +1,165 @@
+		fsWrapper,
+		rootDir,
+	}
+}
+
+//NewGetterFor returns a SourceGetter configured with "url" and "kymaVersion" taken from Installation CR instance.
+func (sgc *sourceGetterCreator) NewGetterFor(legacyKymaSourceConfig LegacyKymaSourceConfig) SourceGetter {
+
+	//In case ALL components have user-defined sources, we don't have to fallback to "defaults" at all. That's why defaultSourcesHandler is invoked lazily with the help of this function.
+	defPkgFn := func() (KymaPackage, error) {
+		res, err := sgc.defaultSourcesHandler.ensureDefaultSources(legacyKymaSourceConfig.KymaURL, legacyKymaSourceConfig.KymaVersion)
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
+	}
+
+	return &componentSrcGetter{
+		fsWrapper:        sgc.fsWrapper,
+		rootDir:          sgc.rootDir,
+		defaultPackageFn: defPkgFn,
+	}
+}
+
+//componentSrcGetter is used to get component sources necessary for the installation
+//Implements SourceGetter
+type componentSrcGetter struct {
+	fsWrapper        FilesystemWrapper
+	rootDir          string
+	defaultPackageFn func() (KymaPackage, error) //Used lazily - only when necessary
+}
+
+//SourceDirFor returns a directory path with charts for given component
+func (csr *componentSrcGetter) SrcDirFor(component v1alpha1.KymaComponent) (string, error) {
+
+	if component.Source != nil && len(component.Source.URL) > 0 {
+		//Handle user-defined component sources
+		log.Printf("Component \"%s\" configured with remote sources", component.Name)
+		return csr.ensureRemoteSourcesFor(component)
+	}
+
+	//Fallback to defaults. These are the sources bundled with the Kyma-operator or downloaded from an external location
+	defaultKymaSources, err := csr.defaultPackageFn()
+	if err != nil {
+		return "", err
+	}
+	var defaultSourcesRoot = defaultKymaSources.GetChartsDirPath()
+	componentDir := path.Join(defaultSourcesRoot, component.Name)
+	return componentDir, nil
+}
+
+func (csr *componentSrcGetter) ensureRemoteSourcesFor(component v1alpha1.KymaComponent) (string, error) {
+	componentSrcLocalCopyDir := path.Join(csr.rootDir, componentsRemoteSrcDir, component.Name)
+	componentChartFile := path.Join(componentSrcLocalCopyDir, "Chart.yaml")
+	if !csr.fsWrapper.Exists(componentChartFile) {
+		log.Printf("Remote sources for component \"%s\" do not exist", component.Name)
+		return csr.getSourcesFor(component, componentSrcLocalCopyDir)
+	}
+
+	log.Printf("Remote sources for component \"%s\" already cached, reusing.", component.Name)
+	return componentSrcLocalCopyDir, nil
+}
+
+func (csr *componentSrcGetter) getSourcesFor(component v1alpha1.KymaComponent, destDir string) (string, error) {
+
+	log.Printf("Fetching sources for component \"%s\" from remote location", component.Name)
+	pwd, err := os.Getwd()
+	if err != nil {
+		return "", err
+	}
+
+	ctx := context.Background()
+	src := component.Source.URL
+	dst := destDir
+	mode := getter.ClientModeDir
+	opts := []getter.ClientOption{}
+
+	var client = &getter.Client{
+		Ctx:     ctx,
```

Comment:
```suggestion
		Ctx:     context.Background(),
```
We could remove some extra lines here